tests/contracts/UserOpGetters.sol

```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.25;
+
+import "@account-abstraction/contracts/interfaces/IEntryPoint.sol";
+
+library UserOpGetters {
+    function getFactory(PackedUserOperation memory op) internal pure returns (address) {
+        if (op.initCode.length < 20) {
+            return address(0);
+        }
+        return bytesToAddress(op.initCode);
+    }
+
+    function getPaymaster(PackedUserOperation memory op) internal pure returns (address) {
+        if (op.paymasterAndData.length < 20) {
+            return address(0);
+        }
+        return bytesToAddress(op.paymasterAndData);
+    }
+
+    function bytesToAddress(bytes memory data) private pure returns (address) {
```

Comment:
that's good thing to know! I thought any kind of slicing is possible only with calldata. so at least prefix is possible with memory too...